packages/runtime/src/binding/ast.ts

```
@@ -154,8 +154,8 @@ export function isLiteral(expr: IsExpressionOrStatement): expr is IsLiteral {
   return (expr.$kind & ExpressionKind.IsLiteral) === ExpressionKind.IsLiteral;
 }
 export function arePureLiterals(expressions: ReadonlyArray<IsExpressionOrStatement>): expressions is IsLiteral[] {
-  const len = expressions && expressions.length || 0;
-  if (len === 0) {
+  const len = expressions && expressions.length;
+  if (len === undefined || len === 0) {
```

@BBosman Gave this a bit more thought and I think, for future reference, that I would actually prefer this:
```ts
if (expressions === undefined || expressions.length === 0) {
    return true;
}
```
Or in this particular instance, perhaps even this:
```ts
if (expressions.length === 0) {
    return true;
}
```
I'd like to move away from all coercions at some point and only use strict equality. If the number of necessary checks become too many, we could further constrain the preconditions (documented at the API boundary with JSDoc comments) and simply let it fail if these are not met.

Reason is to reduce the number of silent failures and enforce more strict pre/postconditions and invariants across the code base, and make sure nothing works or fails by accident and everything is intentional.

In any case, that doesn't change anything for this PR. This is fine for now. I just wanted to point it out

---

@fkleuver In that case can't we just drop the entire `if` here?

The `for` loop will do nothing as the array is empty and then still `return true;`.

---

Good question. It's a consistency and a performance thing. In many situations the method may do additional processing and/or return something different based on whether there's stuff to process or not. Even when that stuff has no effect, preventing potential additional checks and such by returning early is a perf gain.

This is often how the specs operate as well "if nothing to process, return early", even when it could technically let the loop achieve the same behavior. It makes more explicit that there is absolutely nothing to do if there are no items.

It makes no difference here - there would be no impact on performance or behavior, so you're technically right. But I think letting the loop do the short circuit is not necessarily intuitive (especially when debugging) so I'm on the side of how the specs tend to do it.

I don't feel particularly strongly about it however, I just think doing it like this improves code clarity without being a trade-off.

---

Makes sense. 👍